update readme (#272)
FirebaseExtended/firebase-framework-tools: README.md

# App Hosting adapters

## Overview

> App Hosting provides no-config-needed build and deploy support for Web apps developed in these frameworks:

Reviewer: I was wondering where we would land on a term for that . . . I agree that one hyphen is preferable to two.  Right now I don't have any better ideas, so +1!

* Next.js 13+
* Angular 17.2+

> This repo holds the code for the adapters that enable support for these frameworks. At a high level these adapters transform frameowkr specific configurations into an output bundle spec that App Hosting is able to process and configure frameworks support. For more information see our [documentation](https://firebase.google.com/docs/app-hosting/about-app-hosting#frameworks)

Reviewer: One thing I'd add to Daniel's tech-writerly review is: we should use actual target heading text for the link text whenever practical, and though it sounds abrupt, most TWs agree that just "See <link>X</link>" is preferable.

I think we settled on "Framework integration" as a replacement for what is currently "Framework support," so we'd have just "See [Framework integration] (https://firebase.google.com/docs/app-hosting/about-app-hosting#frameworks)."

> This repo holds the code for the adapters that enable support for these frameworks. At a high level these adapters transform frameowkr specific configurations into an output bundle spec that App Hosting is able to process and configure frameworks support. For more information see our [documentation](https://firebase.google.com/docs/app-hosting/about-app-hosting#frameworks)

Reviewer: Typo

Author: I think this may have been an old diff, not sure what happened but these were already addressed for me locally and in the pr I saw. I just pushed another change to address a couple other changes too so could you check again and see if these are fixed now?

> This repo holds the code for the adapters that enable support for these frameworks. At a high level these adapters transform frameowkr specific configurations into an output bundle spec that App Hosting is able to process and configure frameworks support. For more information see our [documentation](https://firebase.google.com/docs/app-hosting/about-app-hosting#frameworks)

Reviewer: I think this should be more like "bundle spec that App Hosting can use to process and configure frameworks support" or just "bundle spec that App Hosting can use to configure frameworks support."

Author: Done!

> This repo holds the code for the adapters that enable support for these frameworks. At a high level these adapters transform framework specific configurations into an [output bundle spec](#app-hosting-output-bundle) that App Hosting can use to configure frameworks support. For more information see our [Framework integration](https://firebase.google.com/docs/app-hosting/about-app-hosting#frameworks).

Reviewer: Nit: omit "our" for just "see [Framework integration]"

## App Hosting output bundle

> The App Hosting output bundle is a file based specification that allows different frameworks to configure and customize their App Hosting deploy to provide more full featured support.

Reviewer: Suggest "deployment"

It feels like a lot of these nits I'm finding were dealt with in a Gdoc, maybe one that Daniel reviewed?  Not sure.

Any framework that can generate a build output in accordance with the App Hosting output bundle can be deployed on App Hosting.

The output bundle primarily consists of a `bundle.yaml` file that sits inside of the `.apphosting` directory. This bundle.yaml contains all the ways that frameworks can configure App Hosting when users deploy their applications.

> NOTE: App Hosting technically supports all all node applications, but no custom framework features will be enabled without the output bundle.

Reviewer 2: You can use this syntax https://github.com/orgs/community/discussions/16925

Author: Done

## Output bundle Schema

> The output bundle all lives inside a single file:

Reviewer: Let's say "is contained in a single file"

Author: Changed!

```shell
.apphosting/bundle.yaml
```

> As long as this file exists and follows the schema App Hosting will be able to process the build properly.

Reviewer: "the schema, App Hosting" (add comma)

Author: Added!

The schema can also be found in <https://github.com/FirebaseExtended/firebase-framework-tools/blob/main/packages/%40apphosting/common/src/index.ts#L4>

```typescript
interface OutputBundle {
  version: "v1"
  runConfig: RunConfig;
  metadata: Metadata;
}
```

### Version

The version represents which output bundle version is currently being used. The current version is v1.

### RunConfig

> The runConfig fields are in charge of configuring the backend’s Cloud run deploy.

Reviewer: If this were Devsite, I'd suggest backticks for all literals, but I'm not 100% what the convention is here.

```typescript
interface RunConfig {
  runCommand: string;
  environmentVariables: EnvVarConfig[];
  concurrency?: number;
  cpu?: number;
  memoryMiB?: number;
  minInstances?: number;
  maxInstances?: number;
}

interface EnvVarConfig {
  variable: string;
  value: string;
  availability: 'RUNTIME'
}

```

| Field  | Type | Description |
| ---------- | ------- | - |

> | runCommand | string |Command to start the server (e.g. "node dist/index.js"). Assume this command is run from the root dir of the workspace.This should be the productionized version of the server start command. |

Reviewer 3: ```suggestion
| runCommand | string |Command to start the server (e.g. "node dist/index.js"). Assume this command is run from the root dir of the workspace. This should be the productionized version of the server start command. |
```
aside - what does "productionized" mean here

| environmentVariables| EnvVarConfig[] | Environment variables present in the server execution environment.|
| concurrency | number | The maximum number of concurrent requests that each server instance can receive.|

> | cpu | number |The number of CPUs used in a single server instance.. |

Reviewer 3: ```suggestion
| cpu | number |The number of CPUs used in a single server instance. |
```

| `memoryMiB` | `number` | The amount of memory available for a server instance.|
| `minInstance` | `number` |The limit on the minimum number of function instances that may coexist at a given time. |
| `MaxInstance` | `number` | The limit on the maximum number of function instances that may coexist at a given time.|

> | `EnvVarConfig.variable` | `string` |Name of the environment variable |

Reviewer 2: I'd rather see EnvVarConfig broken into it's own table to be consistent

Author: Done

| EnvVarConfig.value | string |Value associated with the environment variable |

> | EnvVarConfig.availability | RUNTIME | Where the variable will be available, for now this will always be RUNTIME |

Reviewer 3: ```suggestion
| EnvVarConfig.availability | RUNTIME | Where the variable will be available. For now this will always be RUNTIME |
```

> Many of these fields are shared with apphosting.yaml as well so see [our documentation](https://firebase.google.com/docs/reference/apphosting/rest/v1beta/projects.locations.backends.builds#runconfig) for additional context.

Reviewer: I'd suggest:

"Many of these fields are shared with apphosting.yaml. See the [runConfig reference documentation](https://firebase.google.com/docs/reference/apphosting/rest/v1beta/projects.locations.backends.builds#runconfig) for additional context."

Reviewer: Am I seeing an outdated diff?  It looks like you did not include the suggested change (as well as various other nits)

### Metadata

```typescript
interface Metadata {
  adapterPackageName: string;
  adapterVersion: string;
  framework: string;
  frameworkVersion?: string;
}

```

| Field  | Type | Description |
| ---------- | ------- | - |
| adapterPackageName | string |Name of the adapter (this should be the npm package name) |
| adapterVersion| string | Version of the adapter|
| framework | string | Name of the framework that is being supported|
| frameworkVersion | string |Version of the framework that is being supported |

> Putting this all together a sample bundle.yaml would look as follows:

Reviewer: Suggest

"Putting this all together in a sample bundle.yaml would look like the following"

or maybe just

"Here is a sample bundle.yaml file putting all this together:"

or really:

"Here is a sample bundle.yaml file demonstrating [what we are trying to demonstrate].

I'll stop now :)

```
> cat .apphosting/bundle.yaml

version: v1
runConfig:
  runCommand:
    - node
    - dist/index.js
  environmentVariables:
    - variable: VAR
      value: 8080
      availability: RUNTIME
  concurrency: 80
  cpu: 2
  memoryMiB: 512
  minInstances: 0
  maxInstances: 14
    
metadata:
  adapterNpmPackageName: npm-name
  adapterVersion: 12.0.0
  frameworkNpmPackageName: framework-name
  adapterVersion: 1.0.0

```

> As long as you have the bundle.yaml in this format, App Hosting will be able to deploy any server side rendered framework.

Reviewer 3: ```suggestion
As long as you have the bundle.yaml in this format, App Hosting will be able to deploy any framework that supports support renders.
```

Reviewer: I think Daniel may have meant "that supports server side rendering."
